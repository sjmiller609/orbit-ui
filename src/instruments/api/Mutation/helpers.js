const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
