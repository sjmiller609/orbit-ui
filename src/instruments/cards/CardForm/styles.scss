@import '../../../styles/mixins.scss';

.formContent {
  margin: auto;
  width: 100%;
  max-width: 34rem;
  padding: 1.5rem 0 3rem;
}
