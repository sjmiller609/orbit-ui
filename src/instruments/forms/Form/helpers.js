// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
