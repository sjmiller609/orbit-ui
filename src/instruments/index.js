// NOTE: Any instruments importing other instruments must come AFTER

export { default as Row } from './flex/Row'
export { default as Box } from './flex/Box'

export { default as Card } from './cards/Card'
export { default as Table } from './cards/Table'
export { default as TableRow } from './cards/Table/TableRow'
export { default as CardForm } from './cards/CardForm'

export { default as Module } from './layouts/Module'
export { default as Header } from './layouts/Header'

export { default as Loading } from './animations/Loading'
export { default as LoadingDots } from './animations/LoadingDots'

export { default as Search } from './forms/Search'
export { default as Form } from './forms/Form'
export { default as TextField } from './forms/TextField'
export { default as Password } from './forms/Password'
export { default as Checkbox } from './forms/Checkbox'

export { default as Link } from './links/Link'
export { default as Button } from './links/Button'

export { default as Icon } from './icons/Icon'
export { default as Logo } from './icons/Logo'

export { default as P } from './text/P'
export { default as Mini } from './text/Mini'
export { default as ShowDate } from './text/ShowDate'
export { default as B } from './text/B'

export { default as H1 } from './text/H1'
export { default as H2 } from './text/H2'
export { default as H3 } from './text/H3'
export { default as H4 } from './text/H4'
export { default as H5 } from './text/H5'

export { default as Menu } from './menus/Menu'
export { default as Level1 } from './menus/Level1'

export { default as Snackbar } from './ui/Snackbar'

export { default as ContextUI } from './context/UI/Get'
export { default as ProviderUI } from './context/UI'

export { default as Query } from './api/Query'
export { default as Mutation } from './api/Mutation'
