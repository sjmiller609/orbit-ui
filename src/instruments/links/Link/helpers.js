// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
