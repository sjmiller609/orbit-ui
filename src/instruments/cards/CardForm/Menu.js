'use strict'
import React from 'react'
import PropTypes from 'prop-types'

import { MenuList, Item } from 'instruments'

import s from './styles.scss'

const Menu = ({ menu }) => {
  return (
    <MenuList className={s.menu}>
      {menu.map(m => (
        <Item key={m.id} to={'#' + m.id}>
          {m.text}
        </Item>
      ))}
    </MenuList>
  )
}

Menu.propTypes = {
  menu: PropTypes.array,
}

export default Menu
