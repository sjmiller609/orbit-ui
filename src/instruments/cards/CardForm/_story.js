import React from 'react'
import { storiesOf } from '@storybook/react'
import { withKnobs, text } from '@storybook/addon-knobs'
import { CardForm } from 'instruments'

const form = {
  save: () => null,
  field: {},
}

storiesOf('Instruments|Cards.Form', module)
  .addDecorator(withKnobs)
  .add('No props', () => <CardForm />)
  .add('w/ props', () => (
    <CardForm
      title={text('Title', 'Configure Alerts')}
      button={{
        save: form.save,
        text: text('Button Text', 'Save'),
      }}
    />
  ))
