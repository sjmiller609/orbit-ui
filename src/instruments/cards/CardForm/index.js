'use strict'
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { Card, Button, Row } from 'instruments'
import s from './styles.scss'

// wrap this component in a form, with children as fields, either title as string, or header element, and save buttons in footer
const CardForm = ({
  children,
  button,
  button2,
  footer,
  title,
  className,
  smallForm,
  id,
}) => {
  // console.log(typeof button.save === 'boolean' && !button.save)
  return (
    <Card
      id={id}
      footer={
        footer ||
        (button && (
          <Row className={s.footer}>
            {button2 && (
              <Button
                onClick={button2.onClick}
                style={button2.style}
                className={s.button2}>
                {button2.text || 'Cancel'}
              </Button>
            )}
            <Button
              disabled={
                (typeof button.save === 'boolean' && !button.save) || false
              }
              submit
              arrow="arrow_darkBg"
              onClick={button.onClick}
              to={button.to}
              style={button.style}>
              {button.text || 'Save'}
            </Button>
          </Row>
        ))
      }
      header={title}
      className={className}>
      <div className={classnames(s.formContent, smallForm && s.small)}>
        {children}
      </div>
    </Card>
  )
}

CardForm.propTypes = {
  title: PropTypes.string,
  id: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.element])
    .isRequired,
  footer: PropTypes.oneOfType([PropTypes.array, PropTypes.element]),
  button: PropTypes.object,
  button2: PropTypes.object,
  className: PropTypes.string,
  smallForm: PropTypes.bool,
}

export default CardForm
