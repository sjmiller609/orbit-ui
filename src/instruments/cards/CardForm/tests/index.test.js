import React from 'react'
import { shallow } from 'enzyme'
import { mountWrap } from 'helpers/tests'
import { Card } from 'instruments'
import CardForm from '../index'

const props = {
  title: 'Test',
  id: '123',
  footer: <p>Footer</p>,
  button: {
    onClick: jest.fn(),
  },
  button2: {
    onClick: jest.fn(),
  },
  className: 'testClassString',
  smallForm: false,
  disable: false,
}

describe('CardForm', () => {
  let wrapper
  let card

  beforeEach(() => {
    wrapper = mountWrap(
      <CardForm {...props}>
        <div>Child</div>
      </CardForm>
    )
    card = shallow(
      <Card>
        <div>Child</div>
      </Card>
    )
  })

  it('should be defined', () => {
    expect(CardForm).toBeDefined()
    expect(Card).toBeDefined()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
    expect(card).toMatchSnapshot()
  })

  it('should should no buttons', () => {
    wrapper = mountWrap(
      <CardForm {...props} button={null} button2={null}>
        <div>Child</div>
      </CardForm>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should should render small', () => {
    wrapper = mountWrap(
      <CardForm {...props} smallForm={true}>
        <div>Child</div>
      </CardForm>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should should render disabled', () => {
    wrapper = mountWrap(
      <CardForm {...props} disable={true} button={{ save: false }}>
        <div>Child</div>
      </CardForm>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should should render disabled', () => {
    wrapper = mountWrap(
      <CardForm {...props} disable={true} button={{ save: true }}>
        <div>Child</div>
      </CardForm>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should should render without footer', () => {
    wrapper = mountWrap(
      <CardForm {...props} footer={null}>
        <div>Child</div>
      </CardForm>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
