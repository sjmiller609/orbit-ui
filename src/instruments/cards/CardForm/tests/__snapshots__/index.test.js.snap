// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardForm should render correctly 1`] = `
<CardForm
  button={
    Object {
      "onClick": [Function],
    }
  }
  button2={
    Object {
      "onClick": [Function],
    }
  }
  className="testClassString"
  disable={false}
  footer={
    <p>
      Footer
    </p>
  }
  id="123"
  smallForm={false}
  title="Test"
>
  <Card
    className="testClassString"
    footer={
      <p>
        Footer
      </p>
    }
    header="Test"
    id="123"
  >
    <Box
      className="card testClassString"
      id="123"
      justify="space-between"
    >
      <Flex
        className="card testClassString"
        flow="box"
        id="123"
        justify="space-between"
      >
        <div
          className="flex box space-between card testClassString"
          id="123"
        >
          <header>
            <Box
              className="title"
            >
              <Flex
                className="title"
                flow="box"
              >
                <div
                  className="flex box undefined title"
                >
                  <H5>
                    <h5
                      className="h5"
                    >
                      Test
                    </h5>
                  </H5>
                </div>
              </Flex>
            </Box>
          </header>
          <main
            className="full"
          >
            <div
              className="formContent"
            >
              <div>
                Child
              </div>
            </div>
          </main>
          <footer>
            <p>
              Footer
            </p>
          </footer>
        </div>
      </Flex>
    </Box>
  </Card>
</CardForm>
`;

exports[`CardForm should render correctly 2`] = `
<Box
  className="card"
  justify="flex-start"
>
  <main
    className="full"
  >
    <div>
      Child
    </div>
  </main>
</Box>
`;

exports[`CardForm should should no buttons 1`] = `
<CardForm
  button={null}
  button2={null}
  className="testClassString"
  disable={false}
  footer={
    <p>
      Footer
    </p>
  }
  id="123"
  smallForm={false}
  title="Test"
>
  <Card
    className="testClassString"
    footer={
      <p>
        Footer
      </p>
    }
    header="Test"
    id="123"
  >
    <Box
      className="card testClassString"
      id="123"
      justify="space-between"
    >
      <Flex
        className="card testClassString"
        flow="box"
        id="123"
        justify="space-between"
      >
        <div
          className="flex box space-between card testClassString"
          id="123"
        >
          <header>
            <Box
              className="title"
            >
              <Flex
                className="title"
                flow="box"
              >
                <div
                  className="flex box undefined title"
                >
                  <H5>
                    <h5
                      className="h5"
                    >
                      Test
                    </h5>
                  </H5>
                </div>
              </Flex>
            </Box>
          </header>
          <main
            className="full"
          >
            <div
              className="formContent"
            >
              <div>
                Child
              </div>
            </div>
          </main>
          <footer>
            <p>
              Footer
            </p>
          </footer>
        </div>
      </Flex>
    </Box>
  </Card>
</CardForm>
`;

exports[`CardForm should should render disabled 1`] = `
<CardForm
  button={
    Object {
      "save": false,
    }
  }
  button2={
    Object {
      "onClick": [Function],
    }
  }
  className="testClassString"
  disable={true}
  footer={
    <p>
      Footer
    </p>
  }
  id="123"
  smallForm={false}
  title="Test"
>
  <Card
    className="testClassString"
    footer={
      <p>
        Footer
      </p>
    }
    header="Test"
    id="123"
  >
    <Box
      className="card testClassString"
      id="123"
      justify="space-between"
    >
      <Flex
        className="card testClassString"
        flow="box"
        id="123"
        justify="space-between"
      >
        <div
          className="flex box space-between card testClassString"
          id="123"
        >
          <header>
            <Box
              className="title"
            >
              <Flex
                className="title"
                flow="box"
              >
                <div
                  className="flex box undefined title"
                >
                  <H5>
                    <h5
                      className="h5"
                    >
                      Test
                    </h5>
                  </H5>
                </div>
              </Flex>
            </Box>
          </header>
          <main
            className="full"
          >
            <div
              className="formContent"
            >
              <div>
                Child
              </div>
            </div>
          </main>
          <footer>
            <p>
              Footer
            </p>
          </footer>
        </div>
      </Flex>
    </Box>
  </Card>
</CardForm>
`;

exports[`CardForm should should render disabled 2`] = `
<CardForm
  button={
    Object {
      "save": true,
    }
  }
  button2={
    Object {
      "onClick": [Function],
    }
  }
  className="testClassString"
  disable={true}
  footer={
    <p>
      Footer
    </p>
  }
  id="123"
  smallForm={false}
  title="Test"
>
  <Card
    className="testClassString"
    footer={
      <p>
        Footer
      </p>
    }
    header="Test"
    id="123"
  >
    <Box
      className="card testClassString"
      id="123"
      justify="space-between"
    >
      <Flex
        className="card testClassString"
        flow="box"
        id="123"
        justify="space-between"
      >
        <div
          className="flex box space-between card testClassString"
          id="123"
        >
          <header>
            <Box
              className="title"
            >
              <Flex
                className="title"
                flow="box"
              >
                <div
                  className="flex box undefined title"
                >
                  <H5>
                    <h5
                      className="h5"
                    >
                      Test
                    </h5>
                  </H5>
                </div>
              </Flex>
            </Box>
          </header>
          <main
            className="full"
          >
            <div
              className="formContent"
            >
              <div>
                Child
              </div>
            </div>
          </main>
          <footer>
            <p>
              Footer
            </p>
          </footer>
        </div>
      </Flex>
    </Box>
  </Card>
</CardForm>
`;

exports[`CardForm should should render small 1`] = `
<CardForm
  button={
    Object {
      "onClick": [Function],
    }
  }
  button2={
    Object {
      "onClick": [Function],
    }
  }
  className="testClassString"
  disable={false}
  footer={
    <p>
      Footer
    </p>
  }
  id="123"
  smallForm={true}
  title="Test"
>
  <Card
    className="testClassString"
    footer={
      <p>
        Footer
      </p>
    }
    header="Test"
    id="123"
  >
    <Box
      className="card testClassString"
      id="123"
      justify="space-between"
    >
      <Flex
        className="card testClassString"
        flow="box"
        id="123"
        justify="space-between"
      >
        <div
          className="flex box space-between card testClassString"
          id="123"
        >
          <header>
            <Box
              className="title"
            >
              <Flex
                className="title"
                flow="box"
              >
                <div
                  className="flex box undefined title"
                >
                  <H5>
                    <h5
                      className="h5"
                    >
                      Test
                    </h5>
                  </H5>
                </div>
              </Flex>
            </Box>
          </header>
          <main
            className="full"
          >
            <div
              className="formContent small"
            >
              <div>
                Child
              </div>
            </div>
          </main>
          <footer>
            <p>
              Footer
            </p>
          </footer>
        </div>
      </Flex>
    </Box>
  </Card>
</CardForm>
`;

exports[`CardForm should should render without footer 1`] = `
<CardForm
  button={
    Object {
      "onClick": [Function],
    }
  }
  button2={
    Object {
      "onClick": [Function],
    }
  }
  className="testClassString"
  disable={false}
  footer={null}
  id="123"
  smallForm={false}
  title="Test"
>
  <Card
    className="testClassString"
    footer={
      <Row
        className="footer"
      >
        <Button
          className="button2"
          onClick={[Function]}
          style={undefined}
        >
          Cancel
        </Button>
        <Button
          arrow="arrow_darkBg"
          disabled={false}
          onClick={[Function]}
          style={undefined}
          submit={true}
          to={undefined}
        >
          Save
        </Button>
      </Row>
    }
    header="Test"
    id="123"
  >
    <Box
      className="card testClassString"
      id="123"
      justify="space-between"
    >
      <Flex
        className="card testClassString"
        flow="box"
        id="123"
        justify="space-between"
      >
        <div
          className="flex box space-between card testClassString"
          id="123"
        >
          <header>
            <Box
              className="title"
            >
              <Flex
                className="title"
                flow="box"
              >
                <div
                  className="flex box undefined title"
                >
                  <H5>
                    <h5
                      className="h5"
                    >
                      Test
                    </h5>
                  </H5>
                </div>
              </Flex>
            </Box>
          </header>
          <main
            className="full"
          >
            <div
              className="formContent"
            >
              <div>
                Child
              </div>
            </div>
          </main>
          <footer>
            <Row
              className="footer"
            >
              <Flex
                className="footer"
                flow="row"
              >
                <div
                  className="flex row undefined footer"
                >
                  <Button
                    className="button2"
                    onClick={[Function]}
                  >
                    <Link>
                      <A
                        className="link"
                      >
                        <a
                          className="link"
                        >
                          <button
                            className="button undefined button2"
                            onClick={[Function]}
                            type="button"
                          >
                            Cancel
                          </button>
                        </a>
                      </A>
                    </Link>
                  </Button>
                  <Button
                    arrow="arrow_darkBg"
                    disabled={false}
                    onClick={[Function]}
                    submit={true}
                  >
                    <Link>
                      <A
                        className="link"
                      >
                        <a
                          className="link"
                        >
                          <button
                            className="button undefined"
                            disabled={false}
                            onClick={[Function]}
                            type="submit"
                          >
                            Save
                          </button>
                        </a>
                      </A>
                    </Link>
                  </Button>
                </div>
              </Flex>
            </Row>
          </footer>
        </div>
      </Flex>
    </Box>
  </Card>
</CardForm>
`;
